server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsPhase.java

```diff
@@ -84,7 +88,24 @@ private void hitExecute(Map<String, InnerHitsContext.InnerHitSubContext> innerHi
             if (results == null) {
                 hit.setInnerHits(results = new HashMap<>());
             }
-            innerHitsContext.queryResult().topDocs(topDoc, innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats);
+
+            final DocValueFormat[] sortValueFormats = innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats;
+            List<RescoreContext> rescore = innerHitsContext.rescore();
+            if (rescore.isEmpty() == false) {
+                assert sortValueFormats == null;
+
+                TopDocs justTopDocs = topDoc.topDocs;
+                if (justTopDocs.scoreDocs.length > 0) {
```

Comment:
May be this check is not necessary, as it should be a job of rescorer. Also can `scoreDocs` be null?

Reply:
So it is _technically_ possible for `scoreDocs` to be null, but I highly doubt it ever happens in this code path since the code that follows what I added accesses the `scoreDocs` array without checking if it is null. If it ever was null, we would have seen NPEs in tests by now.